internal/data/metric.go

```diff
@@ -65,42 +77,145 @@ func (md MetricData) MetricCount() int {
+func (md MetricData) ToOtlp() []*otlpmetrics.ResourceMetrics {
+	if !md.pimpl.initialized {
+		// Guaranteed no changes via internal fields.
+		return md.orig
+	}
+	if md.pimpl.sliceChanged {
+		// Reconstruct the slice because we don't know what elements were removed/added.
+		// User may have changed internal fields in any ResourceMetrics, flush all of them.
+		md.orig = make([]*otlpmetrics.ResourceMetrics, len(md.pimpl.resourceMetrics))
+		for i := range md.pimpl.resourceMetrics {
+			md.orig[i] = md.pimpl.resourceMetrics[i].getOrig()
+			md.pimpl.resourceMetrics[i].flushInternal()
+		}
+	} else {
+		// User may have changed internal fields in any ResourceMetrics, flush all of them.
+		for i := range md.pimpl.resourceMetrics {
+			md.pimpl.resourceMetrics[i].flushInternal()
+		}
+	}
+	return md.orig
+}
+
+func (md MetricData) initInternallIfNeeded() {
+	if !md.pimpl.initialized {
+		md.pimpl.resourceMetrics = newResourceMetricsSliceFromOrig(md.orig)
+		md.pimpl.initialized = true
+	}
+}
+
 // ResourceMetrics is a collection of metrics from a Resource.
 //
 // Must use NewResourceMetrics functions to create new instances.
 // Important: zero-initialized instance is not valid for use.
 type ResourceMetrics struct {
-	orig *otlpmetric.ResourceMetrics
+	orig *otlpmetrics.ResourceMetrics
 
 	// Override a few fields. These fields are the source of truth. Their counterparts
 	// stored in corresponding fields of "orig" are ignored.
 	pimpl *internalResourceMetrics
 }
 
 type internalResourceMetrics struct {
-	resource *Resource
+	resource Resource
 	metrics  []Metric
+	// True when the slice was replace.
+	sliceChanged bool
+	// True if the pimpl was initialized.
+	initialized bool
 }
 
 // NewResourceMetrics creates a new ResourceMetrics.
 func NewResourceMetrics() ResourceMetrics {
-	return ResourceMetrics{&otlpmetric.ResourceMetrics{}, &internalResourceMetrics{}}
+	return ResourceMetrics{&otlpmetrics.ResourceMetrics{}, &internalResourceMetrics{}}
+}
+
+// NewResourceMetricsSlice creates a slice of ResourceMetrics that are correctly initialized.
+func NewResourceMetricsSlice(len int) []ResourceMetrics {
+	// Slice for underlying orig.
+	origs := make([]otlpmetrics.ResourceMetrics, len)
+	// Slice for underlying pimpl.
+	pimpls := make([]internalResourceMetrics, len)
+	// Slice for wrappers.
+	wrappers := make([]ResourceMetrics, len)
+	for i := range origs {
+		wrappers[i].orig = &origs[i]
+		wrappers[i].pimpl = &pimpls[i]
+	}
+	return wrappers
+}
+
+func newResourceMetricsSliceFromOrig(origs []*otlpmetrics.ResourceMetrics) []ResourceMetrics {
+	// Slice for underlying pimpl.
+	pimpls := make([]internalResourceMetrics, len(origs))
+	// Slice for wrappers.
+	wrappers := make([]ResourceMetrics, len(origs))
+	for i := range origs {
+		wrappers[i].orig = origs[i]
+		wrappers[i].pimpl = &pimpls[i]
+	}
+	return wrappers
 }
 
-func (rm ResourceMetrics) Resource() *Resource {
+func (rm ResourceMetrics) Resource() Resource {
+	rm.initInternallIfNeeded()
```

Comment:
Can we benchmark this to see how fast are second and subsequent calls? I hope `initInternallIfNeeded` gets inlined. We can look at the generated code at https://go.godbolt.org/
If it is not inlined it may make sense to have `if !rm.pimpl.initialized {` check at every callsite (I know it is more verbose but I am worried function call cost in Go may not be as low as we'd want).

Reply:
Filed issue to track this.